### ISSUE TITLE
[TySan] False positive with struct containing two lists

### ISSUE THREAD

**Issue by tavianator** [llvm/llvm-project#105960]
I am trying out #76261 (commit 733b3ed3f7441453889157834e0a5b6c288bf976).  The following file

```c
struct node {
	struct node *prev, *next;
};

struct list {
	unsigned long n;
	struct node *head, *tail;
};

void list_init(struct list *list) {
	list->n = 0;
	list->head = list->tail = 0;
}

struct lists {
	struct list foo;
	struct list bar;
};

int main(void) {
	struct lists *lists = __builtin_malloc(sizeof(*lists));
	if (lists) {
		list_init(&lists->foo);
		list_init(&lists->bar);
		for (struct node *node = lists->foo.head; node; node = node->next);
	}
	return 0;
}
```

gives a false positive:

```console
tavianator@tachyon $ ~/code/llvm/llvm-project/build/bin/clang -g -fsanitize=type foo.c -o foo
tavianator@tachyon $ ./foo
==478819==ERROR: TypeSanitizer: type-aliasing-violation on address 0x5590bcd3e2a8 (pc 0x559084519178 bp 0x7ffc5c0f0100 sp 0x7ffc5c0f00a8 tid 478819)
READ of size 8 at 0x5590bcd3e2a8 with type any pointer (in lists at offset 8) accesses an existing object of type any pointer (in list at offset 8)
    #0 0x559084519177 in main /home/tavianator/code/bfs/foo.c:25:39
```

**Comment by gbMattN** [comment]
This seems to be the same or a similar problem raised in the TySan discourse (https://discourse.llvm.org/t/reviving-typesanitizer-a-sanitizer-to-catch-type-based-aliasing-violations/66092/23?u=gbmattn), but the current WIP patch does not work for this case.

I'll look into this and see if I can improve the patch!

**Comment by gbMattN** [comment]
! Oh wow! ... Should the commented out line cause a type violation too? Or is everything in the case above (commented line fine, memcpy not) really the correct behaviour?

**Comment by gbMattN** [comment]
Since the fix has been merged, someone with the correct permissions can close this issue out (@goussepi ?)

**Comment by goussepi** [comment]
> Since the fix has been merged, someone with the correct permissions can close this issue out ([@goussepi](https://github.com/goussepi) ?)

Yes I can close.
